### ISSUE TITLE
[Lab 8] NoSuchElementException

### ISSUE THREAD

**Issue by valerietanhx** [nus-cs2030/2223-s1#300]
# Summary

Hello! I'm currently failing the first given test case / test case 8 of Lab 8 due to a NoSuchElementException in my Serve class.

## Description:

I tested the given test case on my working code from before Lab 8 and it throws the same error, so it probably doesn't have anything to do with the self-checkout counters, but I'm at a loss as to what could be causing this issue :( Here's what's going on in each of my events (note that they always return a server list alongside the next Event):

1. Arrive: does nothing, just checks if there's a server that can serve and returns a Serve / Wait / Leave event accordingly
2. Wait: updates the server's queue length, and also just checks if there's a server that can serve. Returns a StillWaiting / Serve event accordingly
3. StillWaiting: basically the same as Wait without the updating queue length part
4. Serve: gets the customer's service time, serves the customer, and updates the queue length
5. Done: lets the server rest, does nothing else
6. Leave: does nothing

Any help would be appreciated! :")

## Code Snippets

Attach code snippets (if any)

## Screenshots

<img width="705" alt="Screenshot 2022-11-24 at 11 32 27 AM" src="https://user-images.githubusercontent.com/110474344/203688265-5dba538f-d3e2-4d81-8760-c90f7b7bad00.png">

**Comment by eLTehh** [comment]
Did you download the new Main from Lab 8? Feels like some variable is missing from the constructor or something.

**Comment by valerietanhx** [comment]
@eLTehh Yup, I did. To try the given test case on my previous code from before Lab 8 I removed the additional self-checkout input. I don't think the problem's related to the constructor, else it wouldn't have compiled.

My current theory is that I'm creating too many Serve events (based on line 6 of the error message), so I'm trying to get a customer's service time too many times, but am not sure where I'm going wrong.

**Comment by mohddanesh** [comment]
the issue seems to be the execute method like what you suggested. Could it be some mistakes in the Wait Event's ifelse statement?

**Comment by izascia256** [comment]
It's hard to tell where the error is from just this. but we can tell the following:

1. Arrival Events are created successfully.
2. Server List is created successfully.

which means the `Simulator::simulate`` method has already added the additional integer for `numOfSelfChecks`.

Since the error appears during `Serve::execute`, you might want to check for what lines of code involves referencing Objects that can carry `null` values (which results in `NoSuchElementException`. i.e. empty `List`, `empty Optionals` , etc.

**Comment by dohaduong** [comment]
Did you maybe try to print the event one by one to see where specifically (which step) it went wrong?

**Comment by Lofthesaver** [comment]
Since the error looks related to the inputs (due to the Scanners), perhaps you can look at whether you are trying to access a customer's service time when it has not yet been served, or when a service time is not needed.

**Comment by wilsonwid** [comment]
I had the exact same error as you, and I think it might be caused due to calling `.get()` on the input times' `Supplier` more than you need to.

**Comment by giorgiocana** [comment]
I bumped into this error a few times. It probably has to do with calling `.get()` method too many times. Thus, just check which code snippets are calling the `.get()` from your supplier, and you'll find the bug.

**Comment by valerietanhx** [comment]
Hi everyone!! Thanks for the tips—figured out that my queue length wasn't updating correctly due to an issue in my Done event, leading to extra customers being served :)

**Comment by Aaronpeh022** [comment]
> Hi everyone!! Thanks for the tips—figured out that my queue length wasn't updating correctly due to an issue in my Done event, leading to extra customers being served :)

Hi, am currently facing the same issue, is there any tips on how you fixed it?

**Comment by acekhoon** [comment]
Oh that actually happened to me as well - took solid 1 hour to debug only to figure out that updating the queues for servers in imlist which is passed onto done class was not updated correctly (i.e. using wrong source to update the queue) such that no such element exception was raised (which was one word change). I think you have to check whether your queue numbers are updated accordingly in your code for each event.

**Comment by Aaronpeh022** [comment]
> Oh that actually happened to me as well - took solid 1 hour to debug only to figure out that updating the queues for servers in imlist which is passed onto done class was not updated correctly (i.e. using wrong source to update the queue) such that no such element exception was raised (which was one word change). I think you have to check whether your queue numbers are updated accordingly in your code for each event.

Yeap, i got it fixed le, thank you so much
